### ISSUE TITLE
DataStore keys should be unicode

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#543]
#480 pt3 was the last straw - attempting to store masters as strings hit the wall of modInfos keys being Path. Path is memory and speed-wise inefficient and we don't want to carry all this mess in py3. Main technical hurdles are:

- settings backwards compat, the main issue here is the TabledInfos.table whose row keys are paths (could drop this for a defauldict maybe but that's later) -> edit on second thought if we store strings there should be ok as bolt.Paths compare with unicode just fine - we can't store pathlib.Path's however
- case insensitive comparisons - some of the data structures should be LowerDicts (or not given that the filesystem won't allow duplicate filenames, and #243 needs a deeper solution)
- caching especially in cext -> maybe time for a new FileInfo.cext attribute (or maybe not if this is not used often/in refresh/in loops)

Blocks #368, #460, #480, #243, #336

**Comment by Utumno** [comment]
Ok kept working on this so some comments - will push some preliminary stuff and the (broken) WIP:

- tracing down Path can be done by tracing GPath or the Path properties
- among those, I targetted `.s`. Due to `Path.__str__` a lot of the .s were superfluous (performance does not matter here it's WIP, and performance is one of the goals). The amount of those .s only shows we actually needed a string
- I was thinking to use CIstr and /or (and) LowerDict to represent paths as one part that GPath'ing masters does is case insensitive comparisons/hashes
- however the more I crawled through uses the more I think it's time for `pathlib`, yep for masters/info keys also. The reason is those are actually treated differently in Unix and co - so we get all the path case-sensitivity correctly handed out of the box _for different operating systems_ - note that:
  - pickling those it's ok as pathlib is in the std
  - we get the str conversion for free (so we do drop .s no prob, gosh that was much work)
  - we get the filename operations for free also and cached
  cons:
  - no caching of pathlib.Path - should be very lightweight but still..
  - case sensitive needs more thought - for instance masters comparison in long formIds. Should we use CIstr here (I intend to cache those - they are immutable)?

So more thought is needed but pushing the .s to nightly for you to have a look - if nothing jump at you (or users) I should merge as this will seriously conflict. I pushed also broken WIP so you have an idea

**Comment by Utumno** [comment]
Third's the charm. You can see the attempt at using CIstr in https://github.com/Utumno/wrye-bash/tree/543-xxx-masters-str. This turned out to be messier and messier:

- those keys should be definitely cached - think only of long form ids. CIStr (and its API LowerDict) were not meant to be cached - remember were introduced for BAIN (see also https://github.com/Utumno/wrye-bash/tree/utumno-263-profiling and https://github.com/wrye-bash/wrye-bash/commit/2897a8f44d1137d505760be579812226ea3f4ba3) and we have a potentially huge amount of. those
- same goes for caching of extension and body - in particular cext and cbody were very frequently used and made sense as (cached) properties - os.spitext calls kept proliferating and this was ugly
-  CIStr was overexposed - remember the API is LowerDict

So it became clear that a new type is needed:

- should be a `str`, mostly those are used as strings and this used to make the code look dirty with all those `.s` - `FName(str)`
- it must compare naturally with other `str` - CIStr does not - see also 66abe7ca79d86d003020adfeded8b186bfa13df8
- it is not a Path - apart from validation which is part of #580 (we get those from listdir
 so they should be valid)
- needs  caching - this must be done on instance creation - so this new type is its own factory (previously known as GPath_ no_norm)
- needs property caching - so we can't use `__slots__`

In all this turned out well - apart from some inevitable complexity:

- backwards/forward compat. Took the opportunity to wrap those in functions. There is a moral here - it pays to store std types and convert on load. This would get us rid of backwards compat __for ever__. Think about it :)
- especially patch config we should rework - thanks to functions we can hook this in to same compat mechanism. In particular storing paths in nested keys is a no-no! (see ImportedMods and configChecks/Items and co, which aslo need leverage oredered `dict` instead of storring an array of items for ordering)
- at least for now we save no FNames in pickles (modulo bugs). This gives us time to solve this (again by pickling strings)
- the API is FNDict. This branch is a rabbit hole and FName appears a lot but should be gradually reconfined - some necessary exposure is due to the fact that it's its own factory, but that's good (aka explicit)

Good old Path served us well (and was anyway patched and repatched see #368 and 2897a8f44d1137d505760be579812226ea3f4ba3) but we need a specialized type here. Timings pending still I bet it actually performs better - need to tune EAFP in comparison methods but cf Path getNorm

For the WIP see my fork repo: https://github.com/Utumno/wrye-bash/tree/utumno-543

**Comment by Infernio** [comment]
Looks really good :+1:

**Comment by Utumno** [comment]
Thanks Inf, I put a lot of effort into this. I am going through the commits (most of them will end up merged, gulp) - I think I got forward compat right but we need to be very careful with backwards - other than that this will really unblock the universe, especially will finally permit us to rest #336 (already installer and Fileinfos share ci_key, although the AFile and Listinfo interaction needs work) and #480 (PluginStr and co)

**Comment by Utumno** [comment]
I just logged in for pinning this <3

Of course, this won't make it to the first beta and I am ok if it does not make it to 310 even, but must be merged first thing after - it's a rabbit hole, gravitationally attracts conflicts like crazy and almost any change in the code needs changes in the 543 branch

**Comment by Infernio** [comment]
First tracebacks from nightly (when annealing via BAIN):
```
Traceback (most recent call last):
  File "bash\balt.py", line 1806, in __Execute
  File "bash\basher\installer_links.py", line 468, in Execute
  File "bash\bosh\bain.py", line 2714, in bain_anneal
  File "bash\bosh\bain.py", line 2661, in _remove_restore
  File "bash\bosh\bain.py", line 2685, in _restoreFiles
AttributeError: 'FName' object has no attribute 's'
Traceback (most recent call last):
  File "bash\balt.py", line 1806, in __Execute
  File "bash\basher\installer_links.py", line 468, in Execute
  File "bash\bosh\bain.py", line 2714, in bain_anneal
  File "bash\bosh\bain.py", line 2661, in _remove_restore
  File "bash\bosh\bain.py", line 2685, in _restoreFiles
AttributeError: 'FName' object has no attribute 's'
Traceback (most recent call last):
  File "bash\balt.py", line 1806, in __Execute
  File "bash\balt.py", line 815, in _conversation_wrapper
  File "bash\basher\installers_links.py", line 200, in Execute
  File "bash\bosh\bain.py", line 2714, in bain_anneal
  File "bash\bosh\bain.py", line 2661, in _remove_restore
  File "bash\bosh\bain.py", line 2685, in _restoreFiles
AttributeError: 'FName' object has no attribute 's'
```

**Comment by Utumno** [comment]
Thanks Inf - should be fixed in 46f4f3cc282fbb465f1a879f6701bb2c72838d58

**Comment by Infernio** [comment]
HEAD of nightly has three commits fixing 543 bugs, wasn't sure where to put them.
